### PR TITLE
 fix(PinchInput): remove recognizer when disconnect 

### PR DESCRIPTION
**Pull request by happyhj**
## Issue
#102 

## Details
fix(PinchInput): remove recognizer when disconnect
* Add inputType option to PinchInput as same as PanInput
* Remove recognizer when disconnect PanInput
* set inputType as touch only at PinchInput unit test because of hammer simulator trigger either touch event or pointer event

### CODE REVIEW

File: src/inputType/PanInput.ts [hunk 1]
```diff
@@ -188,7 +192,11 @@ export class PanInput implements IInputType {
 	*/
 	destroy() {
 		this.disconnect();
-		if (this.hammer && this.hammer.recognizers.length === 1) {
+		if (
+			this.hammer &&
+			this.hammer.recognizers.length === 1 &&
+			this.hammer.recognizers[0] === this.panRecognizer
```
Comment (jongmoon) [inline src/inputType/PanInput.ts]: @happyhj Why is it needed to check `panRecognizer` exist although `disconnect` already removes `panRecognizer`.

File: src/inputType/PinchInput.ts [hunk 2]
```diff
@@ -127,50 +140,61 @@ export class PinchInput implements IInputType {
 	*/
 	destroy() {
 		this.disconnect();
-		if (this.hammer && this.hammer.recognizers.length === 1) {
+		if (
+			this.hammer &&
+			this.hammer.recognizers.length === 1 &&
+			this.hammer.recognizers[0] === this.pinchRecognizer
```
Comment (jongmoon) [inline src/inputType/PinchInput.ts]: @happyhj Same as `PanRecognizer`

File: test/unit/inputType/PinchInput.spec.js [hunk 3]
```diff
@@ -49,7 +49,7 @@ describe("PinchInput", () => {
   describe("enable/disable", function() {
     beforeEach(() => {
       this.el = sandbox();
-      this.inst = new PinchInput(this.el);
+      this.inst = new PinchInput(this.el, {inputType: ["touch"]});
```
Comment (jongmoon) [inline test/unit/inputType/PinchInput.spec.js]: @happyhj Is it the reason simulator set input type as touch mode?

Comment (jongmoon) [reply test/unit/inputType/PinchInput.spec.js]: @happyhj Thank you! 👍

File: src/inputType/PanInput.ts [hunk 4]
```diff
@@ -162,18 +165,18 @@ export class PanInput implements IInputType {
 				throw new Error("Wrong inputType parameter!");
 			}
 			this.hammer = createHammer(this.element, { ...{
-				recognizers: [
-					[Hammer.Pan, hammerOption],
-				],
 				inputClass,
 			}, ... this.options.hammerManagerOptions });
+			this.panRecognizer = new Hammer.Pan(hammerOption);
```
Comment (jongmoon) [inline src/inputType/PanInput.ts]: @happyhj How about to add recognizer outside of condition

```ts
if (this.hammer) {
} else {
}
this.panRecognizer = new Hammer.Pan();
this.hammer.add(this.panRecognizer);
// ...
```

File: src/inputType/PinchInput.ts [hunk 5]
```diff
@@ -84,37 +88,45 @@ export class PinchInput implements IInputType {
...
-			}, ...this.options.hammerManagerOptions});
+			this.hammer = createHammer(
+				this.element,
+				{ ...{
+					inputClass,
+				}, ...this.options.hammerManagerOptions}
+			);
 			this.element[UNIQUEKEY] = keyValue;
+			this.pinchRecognizer = new Hammer.Pinch(hammerOption);
+			this.hammer.add(this.pinchRecognizer);
```
Comment (jongmoon) [inline src/inputType/PinchInput.ts]: @happyhj Same as PanInput

File: src/inputType/PinchInput.ts [hunk 6]
```diff
@@ -127,50 +139,60 @@ export class PinchInput implements IInputType {
 	*/
 	destroy() {
 		this.disconnect();
-		if (this.hammer && this.hammer.recognizers.length === 1) {
+		if (
+			this.hammer &&
+			this.hammer.recognizers.length === 0
```
Comment (jongmoon) [inline src/inputType/PinchInput.ts]: @happyhj how about 1 line?

File: test/unit/Axes.spec.js [hunk 7]
```diff
@@ -301,7 +302,52 @@ describe("Axes", function () {
...
+      this.inst.disconnect(pinchInput);
+      this.inst.connect(["x", "otherX"], panInput);
+
+      const prevX = this.inst.get()["x"];
+      const prevY = this.inst.get()["otherX"];
+
+      // When // pan 제스춰 동작 시 동작해야함 
+			Simulator.gestures.pan(target, MOVE_HORIZONTALLY, () => {
+				Simulator.gestures.pan(target, MOVE_VERTICALLY, () => {
+          // Then
```
Comment (jongmoon) [inline test/unit/Axes.spec.js]: Test is gooood!

Indentation looks not beautiful and it's better to use english than korean

image

File: src/inputType/PanInput.ts [hunk 8]
```diff
@@ -188,7 +191,10 @@ export class PanInput implements IInputType {
 	*/
 	destroy() {
 		this.disconnect();
-		if (this.hammer && this.hammer.recognizers.length === 1) {
```
Comment (younkue) [inline src/inputType/PanInput.ts]: Do you need to check the length of the recognizers when call 'destory' method?

before ```this.hammer = null;```, must call 'destory' method

File: src/inputType/PinchInput.ts [hunk 9]
```diff
@@ -127,50 +136,57 @@ export class PinchInput implements IInputType {
 	*/
 	destroy() {
 		this.disconnect();
-		if (this.hammer && this.hammer.recognizers.length === 1) {
+		if (this.hammer && this.hammer.recognizers.length === 0) {
```
Comment (younkue) [inline src/inputType/PinchInput.ts]: same PanInput's review.

File: src/inputType/PanInput.ts (hunk 8 above)
Comment (happyhj) [reply src/inputType/PanInput.ts]: That condition is checking whether the other input type is not using same hammer instance like PinchInput.

**Pull request merged**